test/test_torch.py

```diff
@@ -5993,6 +5993,37 @@ def test_diagonal_multidim(self, device, dtype):
         self.assertEqual(expected.shape, result.shape)
         self.assertEqual(expected, result)
 
+    def _test_trace(self, device, dtype, legacy):
+        def test(shape):
+            tensor = make_tensor(shape, device, dtype, low=-9, high=9)
+            diag = tensor.diag()
+            if legacy:
+                # NB: trace on cpu doesn't do type promotion...
+                expected = tensor.diag().sum().to(dtype)
+            else:
+                expected = tensor.diag().sum()
+            self.assertEqual(tensor.trace(), expected)
```

Comment:
I think it would be nice to compare it to `np.trace()` result here

Reply:
That feels redundant, especially since numpy has slightly different type promotion semantics so we'll end up trying to force the result to the correct dtype

Reply:
isn't our type promotion different from that of numpy only while dealing with `torch.half` dtype? CCing @mruberry since he is working on testing and numpy compatibility

Reply:
Comparing with NumPy and citing the issue would still be preferable. NumPy's trace supports a dtype argument that should make writing the comparison easy.

Reply:
Is there a torch-to-numpy dtype conversion somewhere?

Reply:
This dict and its sister dict (see below it) handle the conversion for every type except bfloat16: 

https://github.com/pytorch/pytorch/blob/c10aa44e337d27b46efc997a16391d927bac8500/torch/testing/_internal/common_utils.py#L359

For bfloat16 you can cast to float32 when running on NumPy.

Reply:
Thanks!